### PR TITLE
Rewrite 'unique username system', removing libbsd dependency

### LINKED ISSUE #72: Multiple players may have the same name.
**Describe the bug**
In the event that any player has a name (depending on player ID length) 15 or 14 characters in length, another player may freely copy that name, with any of the following effects:
 - Nothing to distinguish the two names whatsoever.
 - (70025abf4472ea4781b964477a3e79530b6fdcfd) The full player ID is concatenated, while making the player name too long in the process.
 - A single character of the player ID is concatenated, (70025abf4472ea4781b964477a3e79530b6fdcfd) resulting in a player name which is too long.

Anything with (70025abf4472ea4781b964477a3e79530b6fdcfd) and not (https://github.com/SpadesX/SpadesX/commit/75bf890639e68df60a81d61932cd9b112c7b7379) preceding it is fixed with #71. If the player name is less than 15 or 14 characters, the player which copied the name is correctly distinguished from the original by concatenating its player ID to the name.

**To Reproduce**
1. Add a player to the game with a name which is 15 characters in length.
2. Add another one with the same name.

**Expected behavior**
Player names which are not identical (70025abf4472ea4781b964477a3e79530b6fdcfd) or too large.

**Screenshots**
(70025abf4472ea4781b964477a3e79530b6fdcfd) Player names that are 15 characters in length, with the copied name being one character too long.
image

(https://github.com/SpadesX/SpadesX/commit/75bf890639e68df60a81d61932cd9b112c7b7379) Player names that are 15 characters in length, and therefore indistinguishable.
image

(70025abf4472ea4781b964477a3e79530b6fdcfd/https://github.com/SpadesX/SpadesX/commit/75bf890639e68df60a81d61932cd9b112c7b7379) Properly distinguished player names.
image

**Server (please complete the following information):**
 - OS: Arch (GNU/)Linux
 - Version: Variable, see #71.

### PR DESCRIPTION
**Pull request by lokka30**
The code used to ensure usernames are unique for online players has been largely rewritten in this PR.

The major takeaways are:

- This PR aims to resolve #72; it appears to have no trouble modifying players' usernames to be unique regardless of the length of their desired name or their player ID as a string.

- Removed use case of `libbsd`, as this new version does not use `strlcat`.
    - Some other cleanup may be needed with CMake and documentation to finish removing this now-unused dependency.

- I'm almost certain there is a lot to improve with this code - C strings are my greatest weakness, lol. Please lend a hand in updating the code as you see fit.
    - I'm particularly worried about the use of standard string functions that aren't aware of a max bound they should restrict themselves to. Though, if they're not needed, then they're not needed - I'll leave this to the professionals to decide.

If you see this PR in a 'ready for review' state, it's ready for review + merge if you want. If it's a draft, I'm still making changes, feel free to ask for an update.

Thanks

## To-Do List

- [x] Ready for review / merge

### CODE REVIEW

**Comment by lokka30** [comment]
Just wanted to update that this PR is ready for review & merge. Please check my usage of the string functions quite thoroughly as string manipulation is a major gap in my C knowledge.

**Comment by lokka30** [comment]
I may need to add another commit to update the 'dependencies' documentation, iirc just `README.md`, to reflect that `libbsd` is no longer used/required.

File: Source/Server/Packets/ExistingPlayer.c [hunk 1]
```diff
@@ -109,21 +103,48 @@ void receive_existing_player(server_t* server, player_t* player, stream_t* data)
...
+        if (!found_match) {
+            // non-conflicting name. we're done here.
+            break;
+        }
+
+        new_name[0] = '\0';
+        strcpy(new_name, player->name);
+        
+        char id_str[4];
+        snprintf(id_str, 4, "%d", player->id + 15);
```
Comment (rndtrash) [inline Source/Server/Packets/ExistingPlayer.c:136]: `%d` -> `%u`, I know it's very unlikely to overflow, but if it does then the number with a minus sign would occupy 4 characters and the null terminator won't fit.

File: Source/Server/Packets/ExistingPlayer.c [hunk 2]
```diff
@@ -109,21 +103,48 @@ void receive_existing_player(server_t* server, player_t* player, stream_t* data)
...
+        }
+        strncat(new_name, id_str, id_len);
     }
-    if (count > 0) {
-        char idChar[4];
-        snprintf(idChar, 4, "%d", player->id);
-        strlcat(player->name, idChar, 16);
-    }
+    player->name[0] = '\0';
+    strcpy(player->name, new_name);
```
Comment (rndtrash) [inline Source/Server/Packets/ExistingPlayer.c:147]: Same here, I guess.

File: Source/Server/Packets/ExistingPlayer.c [hunk 3]
```diff
@@ -109,21 +103,48 @@ void receive_existing_player(server_t* server, player_t* player, stream_t* data)
...
+
+    // ensure the player has a unique name.
+    int       found_match = 0;
     player_t *connected_player, *tmp;
-    HASH_ITER(hh, server->players, connected_player, tmp)
-    {
-        if (is_past_join_screen(connected_player) && connected_player->id != player->id) {
-            if (strcmp(player->name, connected_player->name) == 0) {
-                count++;
+    char      new_name[17] = "";
```
Comment (rndtrash) [inline Source/Server/Packets/ExistingPlayer.c:110]: Perhaps replace the `17` with something like `MAX_NAME_LEN + 1`?

**Comment by lokka30** [comment]
Thank you very much @rndtrash and @Haxk20. I believe all of your feedback has been addressed in the three commits I've pushed since your posts. The only thing I'm rather uncomfortable with is the `PLAYER_NAME_STRLEN` constants being defined in two separate files instead of including it from a common header file.

File: Source/Server/Packets/ExistingPlayer.c [hunk 4]
```diff
@@ -109,21 +103,48 @@ void receive_existing_player(server_t* server, player_t* player, stream_t* data)
...
             }
         }
+
+        if (!found_match) {
+            // non-conflicting name. we're done here.
+            break;
+        }
+
+        new_name[0] = '\0';
+        strcpy(new_name, player->name);
```
[the remark this answers is not in the document]
Comment (lokka30) [reply Source/Server/Packets/ExistingPlayer.c]: todo: needs to be tested

**Comment by lokka30** [comment]
Thanks guys. Hopefully fixed the last of the named constant issues [here](https://github.com/SpadesX/SpadesX/pull/86/commits/8bae7ffd431b7620cd809c16f782e024378c7377).

What's left to do:
* Do we update the documentation and build scripts in this PR to address the removal of the libbsd dependency?
* I need to check if the null character set [discussed here](https://github.com/SpadesX/SpadesX/pull/86#discussion_r1867455233) is at all useful.
* Once these are sorted, this will be ready for merge. More than happy for the current code to be reviewed further if needed.

File: Source/Server/Packets/ExistingPlayer.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (lokka30) [reply Source/Server/Packets/ExistingPlayer.c]: resolved in [ca38be8](https://github.com/SpadesX/SpadesX/pull/86/commits/ca38be8669328b7924b0f811332709cf7e4e7dba)

**Comment by lokka30** [comment]
Ready for review and merge again. Considering the amount of commits, we may need to squash this. Thanks 🙂

**Review by rndtrash — APPROVED** [review]

**Review by Haxk20 — APPROVED** [review]

**Comment by lokka30** [comment]
Appreciate your time and assistance guys. Thank you very much

**Pull request merged**